### ISSUE TITLE
Propagate original object to a Transform method

### ISSUE THREAD

**Issue by vladshestel** [FluentValidation/FluentValidation#1533]
Hello, @JeremySkinner!

Sometimes to validate a specific object I need to prepare some "validation context" with prefetched data. 

For example, let's assume that I have a `Postal` object with `DestinationCityId` field and I need to check if the city id exists in my database. The scenario works fine with something like:
```csharp
public class Postal { public int DestinationCityId; }
public class PostalValidationContext { public Postal Postal; public ICitiesLookup Cities; }

public class PostalValidator : AbstractValidator<PostalValidationContext> {
  public Validator() {
    RuleFor(context => context.Postal.DestinationCityId)
      .Must((context, cityId) => context.Cities.Contains(cityId));
  }
}
```

But there is no ability to create inner validation context for collection property rule:
```csharp
public class Request { public Postal[] Postals; }
public class RequestValidationContext { public Request Request; public ICitiesLookup Cities; }

public class RequestValidator : AbstractValidator<RequestValidationContext> {
  public RequestValidator() {
    RuleForEach(context => context.Postals)
      /* Warning: no way :C */
      .Transform((context, postal) => new PostalValidationContext { /* ... */ })
      .SetValidator(new PostalValidator()); 
  }
}
```

The solution is simple: create overload for `Transform` method. Done here:
https://github.com/FluentValidation/FluentValidation/pull/1532

Ready to do any changes in my PR to align the project requirements.

**Comment by JeremySkinner** [comment]
Hi @vladshestel, thanks for this. 

This sounds like a good idea and a reasonable approach. However, adding new interface methods as well as changing the Transformer signature are both breaking changes to the API, so this will need to hold off until we release a new major version (as we try and stick to semantic versioning). The next major version scheduled is 10.0, which I would hope to have ready by the end of the year, so it will be a bit of a wait before this can go in I'm afraid.

We're also planning on introducing generics throughout the API in version 10, so the final solution here will need to look a bit different (for more details on the changes coming in 10.0 see https://github.com/FluentValidation/FluentValidation/issues/1527).

I'll try and have a think to see whether there's a way to do what you're after without introducing a breaking change.

**Comment by JeremySkinner** [comment]
Implemented for 10.0
